src/HDF5.jl

```diff
@@ -536,6 +536,7 @@ function h5read(filename, name::AbstractString; pv...)
     try
         obj = getindex(fid, name; pv...)
         dat = read(obj)
+        close(obj)
```

Comment:
should these actually be in the `finally` block?

Reply:
Technically, they might need a set of nested try-catch blocks: you can't close `obj` if `getindex` failed, but then you do want to close it if `read` fails.

Reply:
Though allowing `obj` to escape the try-catch block if `read` fails is basically what the status quo was — GC or libhdf5's close semantics should eventually take care of the object.

Reply:
I'm not 100% confident we are following the correct close semantics in all our usages of the `try`--`finally` blocks, many of them are legacy, but I believe are mostly correct. Ideally, we could minimize their usage to only essential places, since `try` `finally` blocks are slow calls.

Reply:
Yeah, I haven't thought hard about the `try`-`finally` use yet, but IIUC they're most essential in cases like this were a file handle is being opened and closed. I agree in trying to see if more uses can be removed elsewhere, though.

Reply:
> Though allowing obj to escape the try-catch block if read fails is basically what the status quo was — GC or libhdf5's close semantics should eventually take care of the object.

If `read` fails,  `close(obj)` won't execute, and then we fall back to closing via libhdf5's or GC's close semantics, which were demonstrated to cause some regressions. This is an edge case and so the performance isn't critical.

Reply:
Sorry, yeah, I guess I only implied it, but that's what I meant.